### PR TITLE
First cut at documenting PBuilder/PReader API, including semantics.

### PR DESCRIPTION
**Pull request by jsuereth**
* Document assumed primtiives
* Document "state" relationships between method calls.
* Add documentation to each individual method.

Review by @heathermiller or @phaller .  Note: I don't really know how to make scaladoc pretty.

### CODE REVIEW

File: core/src/main/scala/pickling/PBuilderReader.scala [hunk 1]
```diff
@@ -16,29 +16,175 @@ trait Hintable {
...
+ * A builder of pickled content.  This is a mutable API, intended to be called in certain specific ways.
+ *
+ * Here are a few static rules that all picklers must follow when using this interface.
+ *
+ * 1. You will be given a type hint before any beginEntry() call.
+ * 2. There will be one endEntry() for every beginEntry() call.
+ * 3. There will be one endCollection() for every beginCollection() call.
+ * 4. Every beginCollection()/endCollection() pair will be inside a beginEntry()/endEntry() pair.
+ * 5. Every putElement() call must happen within a beginCollection()/endCollection() block.
+ * 6. Every putField() call must happen within a beginEntry()/endEntry() block.
```
Comment (havocp) [inline core/src/main/scala/pickling/PBuilderReader.scala]: Is there a rule that "putField may not be called inside begin/endCollection"?

File: core/src/main/scala/pickling/PBuilderReader.scala [hunk 2]
```diff
@@ -16,29 +16,175 @@ trait Hintable {
...
+ */
 trait PBuilder extends Hintable {
+  /** Called to denote that an object is about to be serialized.
+    * @param picklee
+    *                The object to be serialized.  This may be a primtiive, in which case
+    *                it can be immediately serialized (or you can wait unitl endEntry is called).
+    * @return
+    *                A pbuilder instance a pickler can use to serialize the picklee, if it's a complex type.
+    */
   def beginEntry(picklee: Any): PBuilder
```
Comment (havocp) [inline core/src/main/scala/pickling/PBuilderReader.scala]: is it mandatory that putField for the entry is called on the returned PBuilder?

File: core/src/main/scala/pickling/PBuilderReader.scala [hunk 3]
```diff
@@ -16,29 +16,175 @@ trait Hintable {
...
+
+  /**
+   * Call this to denote that the given primitive, collection or structure being pickled is completed.
+   */
   def endEntry(): Unit
+
+  /**
+   * Denotes that a collection of elements is about to be pickled.
+   *
+   * Note: This must be called after beginEntry()
```
Comment (havocp) [inline core/src/main/scala/pickling/PBuilderReader.scala]: is this called on the builder that beginEntry returns, or on the original builder?

File: core/src/main/scala/pickling/PBuilderReader.scala [hunk 4]
```diff
@@ -16,29 +16,175 @@ trait Hintable {
...
   def beginCollection(length: Int): PBuilder
+
+  /**
+   * Places the next element in the serialized collection.
+   *
+   * Note: This must be called after beginCollection().
+   * @param pickler  A callback which is passed a pickler able to serialize the item in the collection.
+   * @return  A pickler which can serialize the next element of the collection.
+   */
   def putElement(pickler: PBuilder => Unit): PBuilder
```
Comment (havocp) [inline core/src/main/scala/pickling/PBuilderReader.scala]: so each call has to be on the builder returned from the last... is endCollection called on the returned builder from the last putElement, or on the same builder used to call beginCollection, or on the builder returned from beginCollection?

File: core/src/main/scala/pickling/PBuilderReader.scala [hunk 5]
```diff
@@ -16,29 +16,175 @@ trait Hintable {
...
+ *   - ArrayBoolean
+ *   - ArrayFloat
+ *   - ArrayDouble
+ */
 trait PReader extends Hintable {
+  /** The scala reflection mirror used when we need to do any runtime-reflection based unpickling. */
   def mirror: Mirror
+  /** start reading a pickled value, returning the serialized FastTypeTag[_], or our static type "hint" */
   def beginEntry(): FastTypeTag[_]
+  /** Start reading a pickled value, ignoring any type tag. */
```
Comment (havocp) [inline core/src/main/scala/pickling/PBuilderReader.scala]: what is the returned String?

File: core/src/main/scala/pickling/PBuilderReader.scala (hunk 1 above)
Comment (jsuereth) [inline core/src/main/scala/pickling/PBuilderReader.scala]: probably :)

**Comment by jsuereth** [comment]
Closing this in favor of my upcoming PR.

**Pull request closed without merging**